singer-commons/src/main/java/com/pinterest/singer/loggingaudit/client/AuditEventKafkaSender.java

```diff
@@ -0,0 +1,178 @@
+
+  private final LoggingAuditStage stage;
+  private final String host;
+  private ArrayBlockingQueue<LoggingAuditEvent> queue;
+  private KafkaProducer<byte[], byte[]> kafkaProducer;
+  private TSerializer serializer = new TSerializer();
+  private AtomicBoolean cancelled = new AtomicBoolean(false);
+  private String topic;
+  private String name;
+  private Thread thread;
+  private int dequeueWaitInSeconds = 30;
+  private int stopGracePeriodInSeconds = 300;
+  private int callingThreadSleepInSeconds = 30;
+
+
+  public AuditEventKafkaSender(KafkaSenderConfig config,
+                               ArrayBlockingQueue<LoggingAuditEvent> queue,
+                               LoggingAuditStage stage , String host, String name){
+     this.topic = config.getTopic();
+     this.queue = queue;
+     this.stage = stage;
+     this.host = host;
+     this.name = name;
+     this.kafkaProducer = KafkaUtils.createKafkaProducer(config.getKafkaProducerConfig(), name);
+  }
+
+  @Override
+  public void run() {
+    LoggingAuditEvent event = null;
+    ProducerRecord<byte[], byte[]> record = null;
+    byte[] key = null;
+    byte[] value = null;
+
+     while(!cancelled.get()){
+       try {
+         event = this.queue.poll(dequeueWaitInSeconds, TimeUnit.SECONDS);
+         if (event != null){
+           LoggingAuditHeaders loggingAuditHeaders = event.getLoggingAuditHeaders();
+           try {
+             key = serializer.serialize(loggingAuditHeaders);
+             value = serializer.serialize(event);
+             record = new ProducerRecord<byte[], byte[]>(this.topic, key, value);
+             kafkaProducer.send(record, new Callback() {
+               @Override
+               public void onCompletion(RecordMetadata recordMetadata, Exception e) {
+                 if (e != null) {
+                   LOG.warn("[{}] failed to send LoggingAuditEvent to Kafka.", Thread.currentThread().getName(), e);
+                   OpenTsdbMetricConverter.incr(LoggingAuditClientMetrics.AUDIT_CLIENT_SENDER_KAFKA_EXCEPTION, 1,
+                       "host=" +  host, "stage=" + stage.toString());
+                 } else {
+                   OpenTsdbMetricConverter.incr(LoggingAuditClientMetrics.AUDIT_CLIENT_SENDER_KAFKA_SUCCESS, 1,
+                       "host=" +  host, "stage=" + stage.toString());
+                 }
+               }
+             });
+           }catch(TException e){
+             LOG.warn("[{}] failed to construct ProducerRecord because of serialization exception.", Thread.currentThread().getName(), e);
+             OpenTsdbMetricConverter.incr(LoggingAuditClientMetrics.AUDIT_CLIENT_SENDER_SERIALIZATION_EXCEPTION, 1,
+                 "host=" +  host, "stage=" + stage.toString(), "logName=" + loggingAuditHeaders.getLogName());
+           }
+         }
+       }catch(InterruptedException e){
+         LOG.warn("[{}] got interrupted while polling the queue.", Thread.currentThread().getName(), e);
```

Comment:
Interrupt exception doesn't quit out of the loop so in case of any other kind of thread termination this will be stuck. One possible design choice could be to use while(true) and then use an interrupt to simply quit out of the loop, this will avoid having to manipulate loop control to shutdown the thread.

Reply:
When thread is interrupted, then we just break the while loop and exits.  made the change in the update of the PR